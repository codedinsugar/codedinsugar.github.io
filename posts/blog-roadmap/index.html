<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>Blog Roadmap</title>
    <meta name="description" content="SRE &#43; DEV &#43; OPS">
    <meta name="keywords" content='blog, codedinsugar, hugo'>

    <meta property="og:url" content="https://codedinsugar.com/posts/blog-roadmap/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="Blog Roadmap">
    <meta property="og:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="og:image" content="/images/sugar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="Blog Roadmap">
    <meta name="twitter:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="twitter:domain" content="https://codedinsugar.com/posts/blog-roadmap/">
    <meta property="twitter:url" content="https://codedinsugar.com/posts/blog-roadmap/">
    <meta name="twitter:image" content="/images/sugar.png">

    
    <link rel="canonical" href="https://codedinsugar.com/posts/blog-roadmap/" />

    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="https://codedinsugar.com/css/dark.css">

    <script src="https://codedinsugar.com/js/svg-injector.min.js"></script>
    <script src="https://codedinsugar.com/js/feather-icons.min.js"></script>
    <script src="https://codedinsugar.com/js/main.js"></script>

    
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://codedinsugar.com">
                <img src="https://codedinsugar.com/images/sugar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://codedinsugar.com">codedinsugar</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>Blog Roadmap</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            July 22, 2022
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/blog">blog</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <blockquote>
<p>Before the score comes the goal</p>
</blockquote>
<hr>
<h3 id="july-2022">July 2022</h3>
<ul>
<li>Intro posts</li>
</ul>
<h3 id="august-2022">August 2022</h3>
<ul>
<li>Configure a light|dark|colored mode</li>
</ul>
<h3 id="september-2022">September 2022</h3>
<ul>
<li>Add a custom logo and a main page image</li>
</ul>
<h3 id="october-2022">October 2022</h3>
<ul>
<li>Persist the navigation menu (current styling auto hides it)</li>
</ul>
<h3 id="november-2022">November 2022</h3>
<ul>
<li>Maintain a one post per week minimum standard</li>
</ul>
<h3 id="december-2022">December 2022</h3>
<ul>
<li>End of year wrap up</li>
</ul>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2023 codedinsugar.com</span>
</footer>
</body>
</html>
