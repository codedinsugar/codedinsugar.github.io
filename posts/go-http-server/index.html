<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>secretProject part 3 - http.HandleFunc in Go</title>
    <meta name="description" content="SRE &#43; DEV &#43; OPS">
    <meta name="keywords" content='blog, codedinsugar, hugo, dev, go'>

    <meta property="og:url" content="https://codedinsugar.com/posts/go-http-server/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="secretProject part 3 - http.HandleFunc in Go">
    <meta property="og:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="og:image" content="/images/sugar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="secretProject part 3 - http.HandleFunc in Go">
    <meta name="twitter:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="twitter:domain" content="https://codedinsugar.com/posts/go-http-server/">
    <meta property="twitter:url" content="https://codedinsugar.com/posts/go-http-server/">
    <meta name="twitter:image" content="/images/sugar.png">

    
    <link rel="canonical" href="https://codedinsugar.com/posts/go-http-server/" />

    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="https://codedinsugar.com/css/dark.css">

    <script src="https://codedinsugar.com/js/svg-injector.min.js"></script>
    <script src="https://codedinsugar.com/js/feather-icons.min.js"></script>
    <script src="https://codedinsugar.com/js/main.js"></script>

    
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://codedinsugar.com">
                <img src="https://codedinsugar.com/images/sugar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://codedinsugar.com">codedinsugar</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>secretProject part 3 - http.HandleFunc in Go</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            November 1, 2022
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/dev">dev</a></li>
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/go">go</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <blockquote>
<p>Oh, what a tangled web we weave, when first we practice to deceive!!.</p>
<p>-Sir Walter Scott</p>
</blockquote>
<hr>
<h1 id="previously-on-codedinsugarcom">Previously on codedinsugar.com</h1>
<p>Now that we have our secretProject <a href="https://codedinsugar.com/posts/kubernetes-cluster-pt-2/">Kubernetes cluster built</a>, let&rsquo;s throw some custom apps in there!</p>
<h1 id="keep-it-simple-and-sweet">Keep it simple and sweet</h1>
<p>This is NOT a Kubernetes (k8s) tutorial!  The k8s ecosystem is massive if you&rsquo;re into container orchestration, read their docs <a href="https://kubernetes.io/docs/home/">here</a>.  I use k8s everyday at work and at home and even though it&rsquo;s almost become second nature for me, I still don&rsquo;t know everything about it&hellip;</p>
<p>The purpose of this post is to:</p>
<ul>
<li>Build a simple http server in Go using only the standard library</li>
<li>Deploy that app to k8s</li>
</ul>
<h1 id="listen-serve-handle">Listen, Serve, Handle</h1>
<p>The end result of our http server is this:</p>
<pre tabindex="0"><code>// Test with curl -d &#39;data&#39; localhost:9190
package main

import (
	&#34;fmt&#34;
	&#34;io/ioutil&#34;
	&#34;log&#34;
	&#34;net/http&#34;
)

func main() {
	//HandleFunc is a convenience method that creates an http handler to respond to http requests
	http.HandleFunc(&#34;/&#34;, func(res http.ResponseWriter, req *http.Request) {
		data, err := ioutil.ReadAll(req.Body)
		if err != nil {
			http.Error(res, &#34;Uh oh something didn&#39;t go right...&#34;, http.StatusBadRequest)
		} else {
			fmt.Fprintf(res, &#34;Hello %s\n&#34;, data)
			log.Printf(&#34;Data: %s\n&#34;, data)
		}
	})

	http.HandleFunc(&#34;/goodbye&#34;, func(res http.ResponseWriter, req *http.Request) {
		data, err := ioutil.ReadAll(req.Body)
		if err != nil {
			http.Error(res, &#34;Uh oh something didn&#39;t go right...&#34;, http.StatusBadRequest)
		} else {
			fmt.Fprintf(res, &#34;Goodbye %s\n&#34;, data)
			log.Printf(&#34;Data: %s\n&#34;, data)
		}
	})

	http.ListenAndServe(&#34;:9190&#34;, nil) // 9090 is taken by Cockpit UI
}
</code></pre><p>But let&rsquo;s break down what&rsquo;s happening in <code>main()</code> even further&hellip;</p>
<pre tabindex="0"><code>http.HandleFunc(&#34;/&#34;, func(res http.ResponseWriter, req *http.Request)
</code></pre><p>If you traceback the references from <a href="https://pkg.go.dev/net/http#HandleFunc">http.HandleFunc</a> ultimately you&rsquo;ll see that the <a href="https://pkg.go.dev/net/http#ServeMux">ServeMux</a> type is inferred.  ServeMux is an HTTP request multiplexer. It matches the URL of each incoming request against a list of registered patterns and calls the handler for the pattern that most closely matches the URL.  The pattern that we&rsquo;re matching is the first argument of the function which is <code>&quot;/&quot;</code> aka the root path after the base domain.  The second argument is an interface to the <a href="https://pkg.go.dev/net/http#Handler">Handler</a> type with a single method called <code>ServeHTTP</code> that accepts a <a href="https://pkg.go.dev/net/http#ResponseWriter">ResponseWriter</a> interface type and a <a href="https://pkg.go.dev/net/http#Request">Request</a> struct type.  These guys make the request magic happen!</p>
<pre tabindex="0"><code>data, err := ioutil.ReadAll(req.Body)
if err != nil {
	http.Error(res, &#34;Uh oh something didn&#39;t go right...&#34;, http.StatusBadRequest)
} else {
	fmt.Fprintf(res, &#34;Hello %s\n&#34;, data)
	log.Printf(&#34;Data: %s\n&#34;, data)
}
</code></pre><p>So so now we need to handle what&rsquo;s actually in the request.  The <code>.Body</code> method is of interface type <a href="https://pkg.go.dev/io#ReadCloser">ReadCloser</a> which we&rsquo;ll assign to the <code>data</code> and <code>err</code> variable identifiers.  If <code>err</code> is satisfied then the <code>.Error</code> method will use <code>res</code> to return a specific string and the response code in <code>.StatusBadRequest</code>.  Simple enough yeah?</p>
<p>To test different request paths we&rsquo;ll create another handle but with a different signature using the <code>&quot;/goodbye&quot;</code> path.</p>
<pre tabindex="0"><code>http.ListenAndServe(&#34;:9190&#34;, nil)
</code></pre><p>Request and response, check.  Now what?  We have to tell the system running the http server (my local box) what port to actually listen on which is where <a href="https://pkg.go.dev/net/http#Server.ListenAndServe">http.ListenAndServe</a> comes in at TCP port 9190.</p>
<h1 id="run-and-build">Run and Build</h1>
<p>Coolio, let&rsquo;s see this bad boy in action:</p>
<pre tabindex="0"><code>go run main.go
// produces no output

curl -d &#39;codedinsugar.com&#39; localhost:9190
Hello codedinsugar.com
// ran from a different terminal

2022/11/01 17:47:39 Data: codedinsugar.com
// returned from original terminal

---

curl -d &#39;testingGoodbye&#39; localhost:9190/goodbye
Goodbye testingGoodbye
...
2022/11/01 17:48:30 Data: testingGoodbye
</code></pre><p>Beautiful and working as expected.  I wonder what else we can pass to <code>-d</code>?</p>
<pre tabindex="0"><code>curl -d $(curl ipinfo.io/ip) localhost:9190
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    13  100    13    0     0     36      0 --:--:-- --:--:-- --:--:--    36
Hello 78.255.xxx.xxx
...
2022/11/01 17:49:15 Data: 78.255.xxx.xxx
</code></pre><p>:)</p>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2023 codedinsugar.com</span>
</footer>
</body>
</html>
