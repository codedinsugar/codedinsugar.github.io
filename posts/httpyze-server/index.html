<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>From http.server with love</title>
    <meta name="description" content="SRE &#43; DEV &#43; OPS">
    <meta name="keywords" content='blog, codedinsugar, hugo, python'>

    <meta property="og:url" content="https://codedinsugar.com/posts/httpyze-server/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="From http.server with love">
    <meta property="og:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="og:image" content="/images/sugar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="From http.server with love">
    <meta name="twitter:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="twitter:domain" content="https://codedinsugar.com/posts/httpyze-server/">
    <meta property="twitter:url" content="https://codedinsugar.com/posts/httpyze-server/">
    <meta name="twitter:image" content="/images/sugar.png">

    
    <link rel="canonical" href="https://codedinsugar.com/posts/httpyze-server/" />

    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="https://codedinsugar.com/css/dark.css">

    <script src="https://codedinsugar.com/js/svg-injector.min.js"></script>
    <script src="https://codedinsugar.com/js/feather-icons.min.js"></script>
    <script src="https://codedinsugar.com/js/main.js"></script>

    
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://codedinsugar.com">
                <img src="https://codedinsugar.com/images/sugar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://codedinsugar.com">codedinsugar</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>From http.server with love</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            February 19, 2023
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/blog">blog</a></li>
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/python">python</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <blockquote>
<p>Darling Moneypenny, you know I never even look at another woman..</p>
<p>-James Bond</p>
</blockquote>
<p>So we layed out the plan <a href="https://codedinsugar.com/posts/golden/">here</a>, and now we action it.</p>
<h1 id="lets-just-jump-right-into-it">Let&rsquo;s just jump right into it</h1>
<p>Here&rsquo;s what we&rsquo;re working with:</p>
<pre tabindex="0"><code>$ tree
.
├── Dockerfile
├── README.md
└── src
    ├── httpsrv.py
    └── index.html
</code></pre><p>Obviously the fun bits are going to be in <code>httpsrv.py</code>:</p>
<pre tabindex="0"><code>from http.server import HTTPServer, BaseHTTPRequestHandler

listenPort = 9292

class MyServer(BaseHTTPRequestHandler):
    def do_GET(self):        
        if self.path == &#34;/&#34;:
            self.path = &#34;/index.html&#34;
            try:
                open_index = open(self.path[1:]).read()
                self.send_response(200)
                self.send_header(&#34;Content-type&#34;, &#34;text/html&#34;)
            except:
                open_index = &#34;File not found&#34;
                self.send_response(404)            
            self.end_headers()
            self.wfile.write(bytes(open_index, &#34;utf-8&#34;))

httpd = HTTPServer((&#34;0.0.0.0&#34;, listenPort),MyServer)
print(f&#34;Starting http server on port {listenPort}&#34;)
httpd.serve_forever()
</code></pre><p>And now for the breakdown:</p>
<pre tabindex="0"><code>from http.server import HTTPServer, BaseHTTPRequestHandler
</code></pre><p>I initially wanted to build a Python API using just the standard library so naturally <a href="https://docs.python.org/3/library/http.server.html">http.server</a> works well for basic POC.  Here we&rsquo;re importing the <code>HTTPServer</code> class which implements <code>TCPServer</code> class and will be our main http server listening for requests.  Next we use the <code>BaseHTTPRequestHandler</code> class to handle requests and serve responses.</p>
<pre tabindex="0"><code>listenPort = 9292
</code></pre><p>Port <code>8080</code> is so 90&rsquo;s and I actually keep track of which ports I have open for my apps.  When we instatiate our http server, it&rsquo;ll listen in on port <code>9292</code>.</p>
<pre tabindex="0"><code>class MyServer(BaseHTTPRequestHandler):
    def do_GET(self):        
        if self.path == &#34;/&#34;:
            self.path = &#34;/index.html&#34;
            try:
                open_index = open(self.path[1:]).read()
                self.send_response(200)
                self.send_header(&#34;Content-type&#34;, &#34;text/html&#34;)
            except:
                open_index = &#34;File not found&#34;
                self.send_response(404)            
            self.end_headers()
            self.wfile.write(bytes(open_index, &#34;utf-8&#34;))
</code></pre><p>Cool, more than just one-liners:</p>
<pre tabindex="0"><code>class MyServer(BaseHTTPRequestHandler):
</code></pre><p>We define a <code>MyServer</code> class which inherits from <code>BaseHTTPRequestHandler</code></p>
<pre tabindex="0"><code>def do_GET(self):
</code></pre><p>This will be our only HTTP verb/method which and is also our only class method.</p>
<pre tabindex="0"><code>if self.path == &#34;/&#34;:
            self.path = &#34;/index.html&#34;
            try:
                open_index = open(self.path[1:]).read()
                self.send_response(200)
                self.send_header(&#34;Content-type&#34;, &#34;text/html&#34;)
            except:
                open_index = &#34;File not found&#34;
                self.send_response(404)            
            self.end_headers()
            self.wfile.write(bytes(open_index, &#34;utf-8&#34;))
</code></pre><p>And the logic begins:</p>
<ul>
<li>If the <code>.path</code> property of our class equals the root level of our served application then&hellip;</li>
<li>Read our only html doc at <code>/index.html</code> and send a 200 response code to the requestor&hellip;</li>
<li>Otherwise, enjoy a 404 and a smile :)</li>
</ul>
<pre tabindex="0"><code>httpd = HTTPServer((&#34;0.0.0.0&#34;, listenPort),MyServer)
</code></pre><p>And in walks our http server where we&rsquo;ve passed it&rsquo;s <code>server_address</code> at all network interfaces listening on port 9292, and <code>RequestHandlerClass</code> which happens to be <code>MyServer</code></p>
<pre tabindex="0"><code>httpd.serve_forever()
</code></pre><p>Forever is a bit excessive yeah?</p>
<p>Let&rsquo;s see what&rsquo;s in <code>/index.html</code></p>
<pre tabindex="0"><code>&lt;!DOCTYPE html&gt;
&lt;html&gt;
&lt;body&gt;

&lt;h1&gt;Hello World!&lt;/h1&gt;
&lt;p&gt;Greetings from httPyze/src/index.html&lt;/p&gt;

&lt;/body&gt;
&lt;/html&gt; 
</code></pre><p>I know, so original.  At least we know where the doc lives.</p>
<h1 id="so-what-does-prod-look-like">So what does prod look like?</h1>
<pre tabindex="0"><code>$ python3 httpsrv.py 
Starting http server on port 9292
192.168.1.80 - - [20/Feb/2023 08:39:59] &#34;GET / HTTP/1.1&#34; 200 -
</code></pre><p>And from our browser</p>
<p><img src="/httpyze-server/httpyze-firefox-screenshot.png" alt="localhost:9292" title="Hi there :)"></p>
<hr>
<p>This is a good stopping point.  Let&rsquo;s pick up in the next sesh where we&rsquo;ll dockerize with podman and deploy to Kubernetes!</p>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2023 codedinsugar.com</span>
</footer>
</body>
</html>
