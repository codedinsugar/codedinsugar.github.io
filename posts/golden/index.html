<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>httPyze - a Python model for secretProject</title>
    <meta name="description" content="SRE &#43; DEV &#43; OPS">
    <meta name="keywords" content='blog, codedinsugar, hugo, python'>

    <meta property="og:url" content="https://codedinsugar.com/posts/golden/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="httPyze - a Python model for secretProject">
    <meta property="og:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="og:image" content="/images/sugar.png">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="httPyze - a Python model for secretProject">
    <meta name="twitter:description" content="SRE &#43; DEV &#43; OPS">
    <meta property="twitter:domain" content="https://codedinsugar.com/posts/golden/">
    <meta property="twitter:url" content="https://codedinsugar.com/posts/golden/">
    <meta name="twitter:image" content="/images/sugar.png">

    
    <link rel="canonical" href="https://codedinsugar.com/posts/golden/" />

    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="https://codedinsugar.com/css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="https://codedinsugar.com/css/dark.css">

    <script src="https://codedinsugar.com/js/svg-injector.min.js"></script>
    <script src="https://codedinsugar.com/js/feather-icons.min.js"></script>
    <script src="https://codedinsugar.com/js/main.js"></script>

    
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://codedinsugar.com">
                <img src="https://codedinsugar.com/images/sugar.png" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://codedinsugar.com">codedinsugar</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
            </div>
            
            <div class="nav-link">
                <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <span id="dark-theme-toggle-screen-reader-target" class="sr-only"></span>
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <span id="hamburger-menu-toggle-screen-reader-target" class="sr-only">menu</span>
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/posts/"><span data-feather='code'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/whoami"><span data-feather='cpu'></span> Whoami </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://codedinsugar.com/tags/"><span data-feather='tag'></span> Tags </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <span id="dark-theme-toggle-screen-reader-target" class="sr-only">theme</span>
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>httPyze - a Python model for secretProject</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            December 23, 2022
        </p>

        <ul class="post-tags">
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/blog">blog</a></li>
        
            <li class="post-tag"><a href="https://codedinsugar.com/tags/python">python</a></li>
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <blockquote>
<p>Everything has beauty, but not everyone sees it.</p>
<p>-Confucius</p>
</blockquote>
<h1 id="python-how-i-love-thee-let-me-count-the-ways">Python how I love thee, let me count the ways</h1>
<p>In a recent social engagement, the topic of discussion was Python.  Somewhere in the middle of the conversation I said something along the lines of &ldquo;I love Python, been reading it for years, wish I could write more of it everyday&rdquo;.  Then it dawned on me, why don&rsquo;t I write Python everyday?  As a Senior SRE, I have to know about a lot of different technologies and how they provide solutions to problems no matter how simple or complex.  Granted I definitely <em>read</em> code everyday but I don&rsquo;t always get to <em>write</em> code everyday.  Let&rsquo;s change that.</p>
<h1 id="httpyze-meets-secretproject">httPyze meets secretProject</h1>
<p>My original plan for secretProject was to build everything entirely in Go for the simple reason of wanting to learn Go.  But why should I limit it to just one language?  That&rsquo;s not realistic at all.  Let&rsquo;s change that.</p>
<p>To simulate what I do for work everyday the best approach here is to build some microservices in Go and some in Python.  This way I can develop an observability palette of different flavors.</p>
<p>Some initial menu items that come to mind:</p>
<ul>
<li>A simple API using <code>from http.server import HTTPServer, BaseHTTPRequestHandler</code></li>
<li>Using podman to build the image and push it to a private DockerHub registry</li>
<li>Deploy httPyze as a Kubernetes Deployment + nodePort Service</li>
</ul>
<p>Let the fun begin!</p>

        </p>
    </div>

    <div class="prev-next">
        
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2023 codedinsugar.com</span>
</footer>
</body>
</html>
